External method call through indirection cells

vitek-karas opened the pull request:
Use indirection cells instead of thunks for external method call fixups.
The feature is off by default behind a new COMPLUS_ZapNoRWX switch.

vitek-karas committed:
Refactor ZapInfo's get FunctionEntryPoint to allow for indirect calls through cells

This also removes some amount of code duplication.

vitek-karas committed:
ExternalMethodCalls through indirection cells on X86

Changes fragile native images to use indirection cells instead of thunks for external method fixups. For now only implemented on X86 (Windows).
The new behavior is off by default, can be turned on by setting COMPLUS_ZapNoRWX=1 in the environement when running crossgen.

jkotas commented on src/vm/i386/asmhelpers.asm:
```diff
@@ -1595,6 +1598,51 @@ public _ExternalMethodFixupPatchLabel@0
+    ; EAX is return address into the callsite which has "call [cell]"
+    ; subtract 4 to get the address of the call target, which points to the indirection cell.
+    ; From there load the address of the indirection cell.
+    mov         eax, [eax - 4]
```
I think if you kept this as null, it would work as well. The start of ExternalMethodFixupWorker has similar code that disasembles the callsite when the indirection cell is null.

vitek-karas replied:
I actually tried that, but the code in the worker is broken. More specifically the ExternalMethodFrame is broken (the logic in it to compute the return address is wrong). I didn't fully understand what the frames are for, so I picked the solution with the stub doing this. To me it feels cleaner, as we have separate files for each architecture and can potentially even delete the ifdef mess at the beginning of the worker.

jkotas replied:
> the logic in it to compute the return address is wrong

This sounds suspect. If ExternalMethodFrame gets its return address wrong, the stackwalking through it will be broken too - GC will crash if it happens to run when ExternalMethodFrame is on the stack. It may be worth understanding how it is wrong.

jkotas commented on src/zap/zapinfo.cpp:
```diff
@@ -1915,20 +1928,9 @@ void ZapInfo::getFunctionFixedEntryPoint(CORINFO_METHOD_HANDLE   ftn,
-        
-        pResult->accessType   = IAT_PVALUE;
-        pResult->addr         = pImport;
+        embedFunctionEntryImport(ftn, pResult);
```
There is similar problem on non-x86 even for regular calls. The JIT does tail-call optimization for regular calls on non-x86: If you have code like:

```
static void foo()
{
    ... stuff ... 
    bar();
}
```

The JIT will generate code like this:

```
... stuff ...
jmp [bar]
```

You won't be able to disassemble the callsite in this case ... the address needs to be passed in explicitly. Once you hit it on x64, you can disable the tailcall optimizations to workaround it (canTailCall - it is what R2R does) ... but I think it will need to be fixed.

vitek-karas replied:
I don't fully understand what the problem is I think. I think I understand the point about the tail calls, but not how it relates to the code you marked for this comment. That code was there before, I just deduplicated it between the two "get entrypoint" methods.
About the tail call - we do have logic elsewhere to figure out jump/call targets - we basically look at the instruction and decide which one to use. Assuming JIT always uses the same type of call/jump, we should still be able to do that... right?
If not then we can either:
- disable the tails calls as you mention. I'm wondering if doing it globally for everything like R2R does is not too much of an overkill. For this change we could scope it to external methods only...
- Introduce different thunk, which would only server as a way to identify the callee. We would still do indirect call, but starting with it pointing to the thunk, the worker would use that to figure out what is being called and then patch the indirection cell instead (it should be possible to make it such that by knowing the thunk we can figure out the matching indirection for it). 

Not sure which one is better. Disabling tails calls is definitely less complex... but the perf hit might be "interesting enough".

jkotas replied:
> how it relates to the code you marked for this comment

Sorry, the comment should have been at different spot.

> JIT always uses the same type of call/jump, we should still be able to do that... right?

The problem is that you do not get the right callsite to disassembly when the JIT jumps to the method instead of calling it.

> but the perf hit might be "interesting enough".

Yes...

vitek-karas replied:
So what I'm getting is that you would prefer the "thunk" solution. Which there are 2 flavors of:
1) Callsite uses indirection cell, but that points to thunk which contains the additional information (somehow), but when we resolve it we go and patch the indirection cell leaving the thunk as is - so the thunk becomes RX only.
2) Callsite uses direct call to a thunk, which inside uses indirect call to the stub. When comes to patching we still patch the indirection cell, but this whole thing is hidden from the callsite (as in the callsite itself didn't need to change from what it is today). The thunk is again RX only.

jkotas replied:
There is also 3rd option that the JIT will use a agreed upon contract for tailcalls - always pass the indirection in given register. The JITed code for tailcalls would be:
```
...
lea r10, [indirection]
jmp [r10]
```
Either option would work. Your option (1) has best steady state so it is what I would lean towards.

jkotas commented on src/zap/zapinfo.cpp:
```diff
@@ -1859,17 +1860,40 @@ PVOID ZapInfo::embedDirectCall(CORINFO_METHOD_HANDLE ftn,
         }
 
-        pEntryPointOrThunkToEmbed = m_pImage->GetImportTable()->GetExternalMethodThunk(ftn);
+        if (CLRConfig::GetConfigValue(CLRConfig::INTERNAL_ZapNoRWX))
```
The conflig value checks are relatively slow. You may want to do this check once when we are initializating the sections, and then just use null check against the section pointer here.

vitek-karas replied:
Fixed, thanks for the info.

vitek-karas committed:
Optimize usage of config flags.

As per PR feedback, avoid frequent checks of CLRConfig values. In this case I'm using the virutal section pointer for the external method cells as indication whether to use indirection or thunks.

jkotas commented on src/zap/zapinfo.cpp:
```diff
@@ -1859,17 +1860,40 @@ PVOID ZapInfo::embedDirectCall(CORINFO_METHOD_HANDLE ftn,
+        {
+            _ASSERTE(!IsReadyToRunCompilation());
+            pResult->accessType = IAT_PVALUE;
+            pResult->addr = m_pImage->GetImportTable()->GetExternalMethodCell(ftn);
```
The comment should have been here.

vitek-karas committed:
Revert "Optimize usage of config flags."

This reverts commit d871c3a4b4c0ac6c067ebd726915d1c0b94e13a1.

vitek-karas committed:
Revert "ExternalMethodCalls through indirection cells on X86"

This reverts commit 158ceabdeedce2cf93e4c59e18c98d83bbc511ec.

vitek-karas committed:
Fix missing return

vitek-karas committed:
Initial version of X86 external method call cells

vitek-karas committed:
Cleanup and x64 support for external method cells
